crates/ty_python_semantic/src/types/newtype.rs

```diff
@@ -0,0 +1,242 @@
+use std::collections::BTreeSet;
+
+use crate::Db;
+use crate::semantic_index::definition::{Definition, DefinitionKind};
+use crate::types::constraints::ConstraintSet;
+use crate::types::{ClassType, Type, definition_expression_type};
+use ruff_db::parsed::parsed_module;
+use ruff_python_ast as ast;
+
+/// A `typing.NewType` declaration, either from the perspective of the
+/// identity-function-that-acts-like-a-subtype-in-type-expressions returned by the call to
+/// `typing.NewType`, or from the perspective of instances of that subtype returned by the
+/// identity function. For example:
+///
+/// ```py
+/// import typing
+/// Foo = typing.NewType("Foo", int)
+/// x = Foo(42)
+/// ```
+///
+/// The revealed types there are:
+/// - `typing.NewType`: `Type::ClassLiteral(ClassLiteral)` with `KnownClass::NewType`.
+/// - `Foo`: `Type::KnownInstance(KnownInstanceType::NewType(NewType { .. }))`
+/// - `x`: `Type::NewTypeInstance(NewType { .. })`
+#[salsa::interned(debug, heap_size=ruff_memory_usage::heap_size)]
+#[derive(PartialOrd, Ord)]
```

Comment:
nit: could you add a note similar to this to the doc-comment for the struct?

https://github.com/astral-sh/ruff/blob/bff32a41dc440b30764e9414767794e01d25c265/crates/ty_python_semantic/src/types.rs#L481-L485